src/rust/bitbox02-rust/src/hww/api/electrum.rs

```diff
@@ -0,0 +1,100 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::pb;
+use super::Error;
+
+use pb::response::Response;
+
+use bitbox02::keystore;
+
+const HARDENED: u32 = 0x80000000;
+const ELECTRUM_WALLET_ENCRYPTION_KEYPATH_LEVEL_ONE: u32 = 4541509 + HARDENED;
+const ELECTRUM_WALLET_ENCRYPTION_KEYPATH_LEVEL_TWO: u32 = 1112098098 + HARDENED;
+
+/// Returns the electrum wallet encryption xpub.
+/// `keypath` currently needs to be m/4541509'/1112098098'
+/// Note: the result of this is only meant to be used for encryption by Electrum.
+/// The resulting xpub must not be used to derive addresses or to receive coins.
+pub async fn process(
+    pb::ElectrumEncryptionKeyRequest { keypath }: &pb::ElectrumEncryptionKeyRequest,
+) -> Result<Response, Error> {
+    if *keypath
+        != [
+            ELECTRUM_WALLET_ENCRYPTION_KEYPATH_LEVEL_ONE,
+            ELECTRUM_WALLET_ENCRYPTION_KEYPATH_LEVEL_TWO,
+        ]
+    {
+        return Err(Error::InvalidInput);
+    }
+    let xpub = keystore::encode_xpub_at_keypath(keypath, keystore::xpub_type_t::XPUB)
+        .or(Err(Error::InvalidInput))?;
+
+    Ok(Response::ElectrumEncryptionKey(
+        pb::ElectrumEncryptionKeyResponse { key: xpub },
+    ))
+}
+
+#[cfg(test)]
+mod tests {
+    extern crate std;
+    use super::*;
+
+    use crate::bb02_async::block_on;
+    use bitbox02::testing::{mock, Data, MUTEX};
+    use std::boxed::Box;
+
+    #[test]
+    pub fn test_process() {
+        let _guard = MUTEX.lock().unwrap();
+
+        // All good.
+        mock(Data {
+            keystore_encode_xpub_at_keypath: Some(Box::new(|keypath, xpub_type| {
+                assert_eq!(
+                    keypath,
+                    [
+                        ELECTRUM_WALLET_ENCRYPTION_KEYPATH_LEVEL_ONE,
+                        ELECTRUM_WALLET_ENCRYPTION_KEYPATH_LEVEL_TWO,
+                    ]
+                );
+                assert_eq!(xpub_type, keystore::xpub_type_t::XPUB);
+                Ok("<xpub>".into())
+            })),
+            ..Default::default()
+        });
+        assert_eq!(
+            block_on(process(&pb::ElectrumEncryptionKeyRequest {
+                keypath: vec![
+                    ELECTRUM_WALLET_ENCRYPTION_KEYPATH_LEVEL_ONE,
+                    ELECTRUM_WALLET_ENCRYPTION_KEYPATH_LEVEL_TWO
+                ]
+            })),
+            Ok(Response::ElectrumEncryptionKey(
+                pb::ElectrumEncryptionKeyResponse {
+                    key: "<xpub>".into()
+                },
+            ))
+        );
+
+        // Invalid keypath.
+        mock(Default::default());
+        assert_eq!(
+            block_on(process(&pb::ElectrumEncryptionKeyRequest {
+                keypath: vec![ELECTRUM_WALLET_ENCRYPTION_KEYPATH_LEVEL_ONE, 0]
```

Comment:
I think it'd be nice to have another test with a keypath len != 2

Reply:
done